branch-3.4 -- bugfix -- ZOOKEEPER-2894

xoiss opened the pull request:
Fixes https://issues.apache.org/jira/browse/ZOOKEEPER-2894

phunt commented:
Hi @xoiss please update this PR to follow the guidelines in "generate a pull request" here: https://cwiki.apache.org/confluence/display/ZOOKEEPER/HowToContribute - in particular wrt the summary and description text which really helps folks identifying which PRs to prioritize. Thanks!

anmolnar commented:
@xoiss Looks like a very nice catch. Would you mind creating a separate pull request for the master branch that we can commit first and backport to other branches?

anmolnar commented:
Closing, because @xoiss hasn't been responding for a while. Please re-open if you can come back to this.

xoiss commented:
// xoiss is here

hi, all!
unfortunately i don't have enough time to respond promptly
and moreover i don't check the gmail mail box because of... some reason

ok, what we can do with these?

anmolnar commented:
@xoiss Same story, please rebase.

xoiss commented:
Ok... As I remember, it turned out there is bug in the fix proposed by me here in case of multi-threaded build. So, it remedies the single-threaded build, but breaks the multi-threaded.

The cause is the 'assert' in the else part. I don't remember what was the idea to put it in the else part... to guarantee what. Surely, the else part never happens in the single-threaded version. But, it occurs in 100% of cases in the multi-threaded, and process gets SIGABRT. As soon as I've been using only the single-treaded version those times, I did not notice that in September 2017. We noticed it later with the different project that uses multi-threaded.

Ok, we have already debugged this and refixed in our locally patched fork, but I have not updated this request here.

So, I need to rework and rebase.

anmolnar commented:
ping @xoiss

xoiss commented:
Hello!
I have revised the issue description, and also I have fixed and reworked this PR.

The improved issue: https://issues.apache.org/jira/browse/ZOOKEEPER-2894
Essentially, I left it unchanged. I have read it now and I've found it more-or-less clear.
Surely, it's very huge, but if one reads it with attention, I think she-or-he would get the idea.
The only thing, I emphasized subparts, where is the problem description, and where is the proposal.

The renewed PR: https://github.com/apache/zookeeper/pull/1000
// hey, look! it's the number 1000 :)
I have rebased it on the latest `branch-3.4`.
As I can see, it passed all precommit checks.
And the main thing, I've remedied my patch for the case of multi-threaded build.

So, I hope now it comply with the PR-style.

Thank you!
Alex (aka xoiss)

Please, revise my renewed PR: https://github.com/apache/zookeeper/pull/1000

xoiss commented:
pong @anmolnar

anmolnar commented:
Closing because #1000 has been merged.
